proposals/extensions_v2.md

```diff
@@ -0,0 +1,349 @@
+
+```c#
+extension E
+{
+    // All extension members extend the same thing:
+
+    public void M() for SomeType { ... }
+    public int P for SomeType { get { ... } }
+    public static operator+(...) for SomeType { ... }
+    // etc
+}
+
+// Can be translated to:
+
+extension E for SomeType
+{
+    public void M() { ... }
+    public int P { get { ... } }
+    public static operator+(...) { ... }
+}
+```
+
+TODO: Do an ecosystem check on what percentage of existing extensions could use this simpler form.
+
+TODO: It's possible someone might have an extension where almost all extensions extend a single type, and a small handful do something slightly different (perhaps extending by `this ref`).  Would it be beneficial here to *still* allow the extension members to provide a `for-clause` to override that default for that specific member.  For example:
+
+```c#
+extension StringExtensions for string
+{
+    // Lots of normal extension methods on string ...
+
+    // Override here to extend `string?`
+    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
+    {
+    }
+}
+```
+
+It seems like this would be nice to support with little drawback.
+
+## Expansion 2: Optional syntactic components
+
+As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:
+
+```c#
+static class Extensions
+{
+    // Yes, this is legal
+    public static void MakeNonNull([Attr] this ref int? value)
+    {
+        if (value is null)
+            value = 0;
+    }
+}
+```
+
+For this reason, the strawman syntax for this `for clause` is `for parameter`, where `parameter` is the familiar:
+
+```g4
+parameter
+    | attributes? modifiers? type identifier
+    ;
+```
+
+(fortunately, extension methods today don't support a default value for the `this` parameter, so wel don't have to support that).
+
+However, for many extensions no name is really required.  All extension members (except for extension-constructors and extension-static-methods) are conceptually a way to extend `this` with new functionality.  This is so much so the case that we even designed classic extension methods to use the `this` keyword as their designator.  As such, we forsee potentially making the name optional, allowing one to write an extension like so:
+
+```c#
+extension Enumerable
+{
+    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
+        for IEnumerable<TSource> // no name
+        where TResult : struct, INumber<TResult>
+        where TAccumulator : struct, INumber<TAccumulator>
+    {
+        // Use 'this' in here to represent the value being extended
+    }
+}
+```
+
+This would have to come with some default name chosen by the language for the parameter in metadata.  But that never be needed by anyone calling it from a modern compiler.
```

Comment:
The proposal is aimed at compat and to hit that I do think we need to solve the parameter name of the `this`. There will inevitably be some cusotmer that has bet on calling the method in `static` form using named arguments. If we don't have a solution for that then I fear that teams like runtime will have hesitation in migrating to the new syntax. 

I do wonder if we would want to split that out though into a separate proposal. There are a number of other cases where we've desired to allow named arguments to use a different name than the parameter name within the type:

1. For records there is the tension of do you use pascal or camel case for parameter names? For .NET style you want the field to be pascal cased but the parameter  camel cased. 
2. For dictionary expressions one of the comparer designs includes binding to a parameter named `comparer`. Unfortunately there are plenty of examples where that name was not used for the comparer. Have to decide between compromising the design or breaking compat.

Maybe this design is the final straw in saying lets go and solve that problem generally so we can unlock a number of scenarios. Essentially give developers a way of saying 

```csharp
[ParameterName(0, "stringComparer")]
public void M(IComparer<string> comparer) { ... }

M(stringComparer: null); // okay
M(comparer: null); // okay
```

Reply:
> The proposal is aimed at compat and to hit that I do think we need to solve the parameter name of the this. There will inevitably be some cusotmer that has bet on calling the method in static form using named arguments. If we don't have a solution for that then I fear that teams like runtime will have hesitation in migrating to the new syntax.

The proposal allows the name to be specified just as you can todaay.  It's only in the "future enhancements" where we even consider allowing that to be optional.  

My goal is that hte *core* proposal give *perfect* binary compat.  in other words it generates *identical* IL/metadata.

Reply:
> Maybe this design is the final straw in saying lets go and solve that problem generally so we can unlock a number of scenarios. Essentially give developers a way of saying

My preference is that that is a future enhancement.  I think i want to start us with perfect compat, and then add niceties for cases where it either isn't needed, or can be made simpler.

Reply:
I'm confused how this design achieves perfect compat then. Consider the following: 

```csharp
1.M();
E.M(i: 42);

static class E { 
  internal static void M(this int i) { }
}
```

How can I port this to the new model without the ability to name the `this`?

Reply:
you can name the `this`.  That's exactly what the spec says (and even gives examples).  The for-clause is `for parameter`.  so you write this as:

```c#
extension E
{
    internal void M() for int i { /*exact same body*/ }
}
```

Reply:
Ah okay. That works for this then.

Reply:
Here's the section on how to migrate:

> In other words, all existing extension methods drop `static` from their signature, and move their first parameter to a `for-clause` placed within the method header (currently strawmanned as after the parameter list).  Note: the syntax of a `for-clause` is `'for' parameter`, allowing things like a parameter name to be specified.  `parameter` is critical in this design to ensure the classic extension method `this` parameter can always cleanly move.

> The strawman chooses this location as it already cleanly supports clauses, being where the type parameter constraint clauses already go. 

We can write up a migration guide (and we'll have a fixer).  but the steps (i believe) are simply:

1. change `static class` to `extension`.  Keep the type header otherwise the same.
2. for all extension methods in the static class (those that use `this` in the first parameter):
  a. drop 'static'
  c. remove the `this` parameter from the parameter list.
  d. make a `for-clause` with the original `this` parameter (with the `this` keyword dropped)
  e. place the `for-clause` after the existing parameter list, before any constraints

I believe that's the entire algorithm.  

Conversely, when compiling for *instance extension methods*, we will do *everything* (binding, reducing, lowering, emitting) as if it was declared in the original static-extension-method form where the for-clause is the first parameter.  I've tried very hard to make all sides of this very simple, so the new instance extension methods and the classic static extensions are *trivially* convertible.  We need this (imo) to help ensure the perfect source and binary compat.